mava/systems/ppo/sebulba/ff_ippo.py

```diff
@@ -11,3 +11,707 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+import copy
+import threading
+from queue import Queue
+from typing import Any, Dict, List, Sequence, Tuple
+
+import chex
+import flax
+import hydra
+import jax
+import jax.debug
+import jax.numpy as jnp
+import numpy as np
+import optax
+from colorama import Fore, Style
+from flax.core.frozen_dict import FrozenDict
+from omegaconf import DictConfig, OmegaConf
+from optax._src.base import OptState
+from rich.pretty import pprint
+
+from mava.evaluator import get_sebulba_eval_fn as get_eval_fn
+from mava.evaluator import make_ff_eval_act_fn
+from mava.networks import FeedForwardActor as Actor
+from mava.networks import FeedForwardValueNet as Critic
+from mava.systems.ppo.types import LearnerState, OptStates, Params, PPOTransition
+from mava.types import (
+    ActorApply,
+    CriticApply,
+    ExperimentOutput,
+    Observation,
+    SebulbaLearnerFn,
+)
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax_utils import merge_leading_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.sebulba_utils import ParamsSource, Pipeline, RecordTimeTo, ThreadLifetime
+from mava.utils.total_timestep_checker import sebulba_check_total_timesteps
+from mava.utils.training import make_learning_rate
+from mava.wrappers.episode_metrics import get_final_step_metrics
+
+
+def rollout(
+    key: chex.PRNGKey,
+    config: DictConfig,
+    rollout_pipeline: Pipeline,
+    params_source: ParamsSource,
+    apply_fns: Tuple,
+    actor_device_id: int,
+    seeds: List[int],
+    thread_lifetime: ThreadLifetime,
+) -> None:
+    # setup
+    env = environments.make_gym_env(config, config.arch.num_envs)
+    current_actor_device = jax.devices()[actor_device_id]
+    actor_apply_fn, critic_apply_fn = apply_fns
+    num_agents, num_envs = config.system.num_agents, config.arch.num_envs
+
+    # Define the util functions: select action function and prepare data to share it with learner.
+    @jax.jit
+    def get_action_and_value(
+        params: FrozenDict,
+        observation: Observation,
+        key: chex.PRNGKey,
+    ) -> Tuple:
+        """Get action and value."""
+        key, subkey = jax.random.split(key)
+
+        actor_policy = actor_apply_fn(params.actor_params, observation)
+        action = actor_policy.sample(seed=subkey)
+        log_prob = actor_policy.log_prob(action)
+
+        value = critic_apply_fn(params.critic_params, observation).squeeze()
+        return action, log_prob, value, key
+
+    timestep = env.reset(seed=seeds)
+
+    next_dones = jax.tree_util.tree_map(
+        lambda x: jnp.repeat(x, num_agents).reshape(num_envs, -1),
+        timestep.last(),
+    )
```

Comment:
No need for the `tree_map` here, it's something I've been meaning to remove in mava

```suggestion
    next_dones = jnp.repeat(timestep.last(), num_agents).reshape(num_envs, -1)
```